src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -52,18 +52,29 @@
 import java.util.concurrent.Callable;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
 import java.util.stream.Collectors;
 
 import io.confluent.connect.elasticsearch.ElasticsearchSinkConnectorConfig.BehaviorOnMalformedDoc;
 
 import static io.confluent.connect.elasticsearch.ElasticsearchSinkConnectorConfig.FLUSH_TIMEOUT_MS_CONFIG;
 import static io.confluent.connect.elasticsearch.ElasticsearchSinkConnectorConfig.MAX_BUFFERED_RECORDS_CONFIG;
 
+/**
+ * Implementation notes:
+ * <ul>
+ *   <li>Based on Elasticsearch's BulkProcessor, which is responsible for batching based on size
+ *   and linger. It also limits the concurrency (max number of in-flight requests).</li>
+ *   <li>Batches are run asynchronously on a separate thread</li>
+ *   <li>Retries are processed synchronously in the original batch thread</li>
+ * </ul>
+ */
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
```

Comment:
I don't think this description helps to really explain how this client is now performing the behavior. IIUC, the bulk processor was being used to actually execute the requests, but now that's not the case. Instead, the bulk processor is being used to synchronously submit batches (with retry logic) to the executor service. IOW, the bulk processor is basically just building batches and submit requests to the executor service, which actually sends the requests to ES and retries upon any failure. If the retries all fail, then the exception is reported via the atomic reference to an error, which is checked and thrown from a subsequent call to the task's `put(...)` method and that results in failure of the task.

Overall this sounds like a solid approach, but it's odd this is not described even slightly in the code or JavaDoc.

Reply:
Expanded documentation.